### PR TITLE
perf(ext/node): native vectored write for server streams

### PR DESCRIPTION
**Pull request by littledivy**
```
# main
$ ./load_test 10 0.0.0.0 8080 0 0
Using message size of 20 bytes
Running benchmark now...
Msg/sec: 106182.250000
Msg/sec: 110279.750000
^C

# this PR
$ ./load_test 10 0.0.0.0 8080 0 0
Using message size of 20 bytes
Running benchmark now...
Msg/sec: 131632.250000
Msg/sec: 134754.250000
^C
```

### CODE REVIEW

File: ext/http/http_next.rs [hunk 1]
```diff
@@ -1034,6 +1035,34 @@ impl UpgradeStream {
     .try_or_cancel(cancel_handle)
     .await
   }
+
+  async fn write_vectored(
+    self: Rc<Self>,
+    buf1: &[u8],
```
Comment (aapoalas) [inline ext/http/http_next.rs:1041]: **question**: Is there a particular reason for supporting specifically two buffers worth of data?

Comment (littledivy) [reply ext/http/http_next.rs:1041]: Rust's `write_all_vectored` is not stable, this function is a stable version of that but for 2 buffers. We can support variable size once its stable.

File: ext/http/http_next.rs [hunk 2]
```diff
@@ -1048,3 +1077,16 @@ impl Resource for UpgradeStream {
     self.cancel_handle.cancel();
   }
 }
+
+#[op]
+pub async fn op_raw_write_vectored(
```
[the remark this answers is not in the document]
Comment (littledivy) [reply ext/http/http_next.rs:1082]: We could but maybe we should wait for `write_all_vectored` to be stable? https://github.com/rust-lang/rust/issues/70436

This implementation only works on two iovecs, pretty niche for websocket frames (header + data)

**Review by bartlomieju — COMMENTED** [review]
LGTM, but I want more eyes on this. Pinging @mmastrac for additional review.

File: ext/http/http_next.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (mmastrac) [reply ext/http/http_next.rs:1082]: I am OK with this for now -- I'm going to be revisiting streams after the op work and will look to see if there's a way we can create our own vectored interface.

File: ext/http/http_next.rs [hunk 3]
```diff
@@ -1048,3 +1077,16 @@ impl Resource for UpgradeStream {
...
+  state: Rc<RefCell<OpState>>,
+  rid: ResourceId,
+  buf1: JsBuffer,
+  buf2: JsBuffer,
+) -> Result<usize, AnyError> {
+  let resource: Rc<UpgradeStream> =
```
Comment (mmastrac) [inline ext/http/http_next.rs:1088]: If this is not an `UpgradeStream`, it should probably fall back to the standard write methods. I don't think we should assume that any two-chunk write is a websocket write.

Comment (littledivy) [reply ext/http/http_next.rs:1088]: In JS, All `TCPSERVERWRAP` provider type resources are an UpgradeStream so this does not need a fallback.

File: ext/node/polyfills/internal_binding/stream_wrap.ts [hunk 4]
```diff
@@ -195,6 +198,31 @@ export class LibuvStreamWrap extends HandleWrap {
...
     allBuffers: boolean,
   ): number {
+    const supportsWritev = this.provider === providerType.TCPSERVERWRAP;
+    // Fast case optimization: two chunks, and all buffers.
+    if (chunks.length === 2 && allBuffers && supportsWritev) {
+      // String chunks.
+      if (typeof chunks[0] === "string") chunks[0] = Buffer.from(chunks[0]);
+      if (typeof chunks[1] === "string") chunks[1] = Buffer.from(chunks[1]);
+
+      ops.op_raw_write_vectored(
```
[the remark this answers is not in the document]
Comment (littledivy) [reply ext/node/polyfills/internal_binding/stream_wrap.ts:208]: `chunks.length == 1` should be using `write` instead of `writev`

**Review by mmastrac — APPROVED** [review]
LGTM. Let's land for now and work on vectorizing resources

**Pull request merged**